src/composition.jl

```diff
@@ -1,18 +1,18 @@
-struct CompositeMap{T, As<:Tuple{Vararg{LinearMap}}} <: LinearMap{T}
+struct CompositeMap{T, As<:LinearMapTuple} <: LinearMap{T}
     maps::As # stored in order of application to vector
     function CompositeMap{T, As}(maps::As) where {T, As}
         N = length(maps)
         for n in 2:N
             check_dim_mul(maps[n], maps[n-1])
         end
-        for n in eachindex(maps)
-            A = maps[n]
-            @assert promote_type(T, eltype(A)) == T "eltype $(eltype(A)) cannot be promoted to $T in CompositeMap constructor"
+        for TA in Base.Generator(eltype, maps)
+            promote_type(T, TA) == T ||
+                error("eltype $TA cannot be promoted to $T in CompositeMap constructor")
```

Comment:
That's clever! That makes the loop body depend on types, and avoids closure about `maps`, right?

Reply:
if @dkarrasch has to ask about this, then others like me will have even less chance of understanding so this would be a nice place to add a comment in the code...

Reply:
In Julia 1.x there is now `Base.Iterators.map`, a lazy map, which does exactly this, it just calls `Base.Generator`. `Base.Generator` is what is being called by the generator syntax. It never creates a closure over the iterator, only over other arguments provided in the function. Say you have somewhere
```
f(x, a) for x in b
```
that basically lowers to `Base.Generator(x->f(x,a), b)` and thus creates a closure over/of/about (what is the correct preposition?) `a`.

Reply:
I will add a comment.